internal/resourceproviders/required.go

```diff
@@ -3,22 +3,131 @@
 
 package resourceproviders
 
-// Required returns core Resource Providers used by the AzureRM Provider
-// Terraform auto-registers core Resource Providers, since those RP’s should be enabled by default
-// but that list is something we come up with based on experience.
-// whilst all may not be used by every user - the intention is that we determine which should be
-// registered such that we can avoid obscure errors where Resource Providers aren't registered.
-// new core Resource Providers should be added to this list as they're used in the Provider
-// (this is the approach used by Microsoft in their tooling)
-func Required() map[string]struct{} {
-	// NOTE: Resource Providers in this list are case sensitive
-	return map[string]struct{}{
-		"Microsoft.AppConfiguration":        {},
+// This file contains sets of resource providers which the provider should automatically register, depending on
+// configuration by the user. Historically, we have ordained the same set of RPs for all users, and users could
+// enable or disable automatic registration of the RPs in that set.
+//
+// Users may now choose a set of RPs that best meets their needs, via the provider block, or they can provide their
+// own set. They can also choose to not register any RPs, and manage these out-of-band, if they wish.
+//
+// Note that resource providers are case-sensitive.
+//
+// Official Docs: https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers
+
+type ResourceProviders map[string]struct{}
+
+func (r ResourceProviders) Add(providers ...string) {
+	for _, p := range providers {
+		r[p] = struct{}{}
+	}
+}
+
+func (r ResourceProviders) Merge(a ResourceProviders) ResourceProviders {
+	for k, v := range a {
+		r[k] = v
+	}
+	return r
+}
+
+// Core is a minimal set of RPs, and will be the default set in a future major version of the provider.
+func Core() ResourceProviders {
+	return ResourceProviders{
+		"Microsoft.Authorization":       {},
+		"Microsoft.Compute":             {},
+		"Microsoft.CostManagement":      {},
+		"Microsoft.ManagedIdentity":     {},
+		"Microsoft.MarketplaceOrdering": {},
+		"Microsoft.Network":             {},
+		"Microsoft.Resources":           {},
+		"Microsoft.Storage":             {},
+	}
+}
+
+// Extended is an expanded set of resource providers, as suggested by the community.
+func Extended() ResourceProviders {
+	return Core().Merge(ResourceProviders{
+		"Microsoft.ApiManagement":        {},
+		"Microsoft.AppConfiguration":     {},
+		"Microsoft.AppPlatform":          {},
+		"Microsoft.Automation":           {},
+		"Microsoft.Cache":                {},
+		"Microsoft.Cdn":                  {},
+		"Microsoft.ContainerInstance":    {},
+		"Microsoft.ContainerRegistry":    {},
+		"Microsoft.ContainerService":     {},
+		"Microsoft.DBforMariaDB":         {},
+		"Microsoft.DBforMySQL":           {},
```

Comment:
We've deprecated and will remove all of these resources in 4.0, so might not be necessary to add their RPs here 
```suggestion
```

Reply:
Cool thanks, will remove these 👍